### PR TITLE
ci: introduce GHA workflow to test BPF tproxy with Ingress

### PR DESCRIPTION
**Pull request by mhofstetter**
Currently, we're lacking test coverage for BPF tproxy to redirect traffic transparently to the L7 Proxy (e.g. Envoy).

Therefore, this commit introduces an additional GitHub action workflow that configures `bpf.tproxy=true` and executes a minimal set of ingress related connecitivy tests.

Related issue: https://github.com/cilium/cilium/issues/30331

### CODE REVIEW

**Comment by mhofstetter** [comment]
Currently, the new check is failing as expected: https://github.com/cilium/cilium/actions/runs/7644144588/job/20827778279?pr=30404

Extract from the sysdump's hubble flow

```
~/sysdump_tproxy_test/cilium-sysdump-20240124-175145
❯ hubble observe --input-file hubble-flows-cilium-qc752-20240124-175145.json  | grep DROPPED
...
Jan 24 17:51:43.613: cilium-test/client-69748f45d8-68p56:33372 (ID:29224) <> cilium-test/cilium-ingress-ingress-service:80 (world) Socket lookup failed DROPPED (TCP Flags: SYN)
Jan 24 17:51:44.626: cilium-test/client-69748f45d8-68p56:33372 (ID:29224) <> cilium-test/cilium-ingress-ingress-service:80 (world) Socket lookup failed DROPPED (TCP Flags: SYN)
...
```

This should pass once the PR with the fix gets merged:  https://github.com/cilium/cilium/pull/30397 is merged

**Review by sayboras — APPROVED** [review]
Nice 👍 

Just a note for other reviewer, only Ingress related tests are run.

File: .github/workflows/conformance-bpf-tproxy.yaml
Comment (tklauser) [inline .github/workflows/conformance-bpf-tproxy.yaml]: I looks like this workflow is triggerd on PR and push and not via Ariane. So the trigger phrase is not needed, unless we want to change the workflow to be triggered by Ariane.

Comment (mhofstetter) [reply .github/workflows/conformance-bpf-tproxy.yaml]: Good catch - thanks! Indeed it was initially an ariane based workflow :)

changed

**Comment by mhofstetter** [comment]
incorporated Tobias feedback (removing trigger in name of workflow)

**Comment by mhofstetter** [comment]
rebased to `main` to pick up fix :crossed_fingers: :pray:

**Comment by mhofstetter** [comment]
worfklow succeeds :heavy_check_mark: : https://github.com/cilium/cilium/actions/runs/7653182352/job/20854498693?pr=30404

File: .github/workflows/conformance-bpf-tproxy.yaml
Comment (joestringer) [inline .github/workflows/conformance-bpf-tproxy.yaml:1]: Thanks for the proposal. I think it makes sense to ensure that this flag is covered in CI to prevent further regressions.

At the same time, I'm a little worried about the permutations we'll end up with if we are creating individual github workflows for highly specific details like bpf.tproxy. From recollection this changes maybe a hundred lines or so of code in the bpf datapath. Do we need an entire workflow run to validate this, or should we just enable this setting in some of the other more advanced workflows?

Maybe to sketch out an idea, we could have a workflow that enables almost everything with an iptables mode, then another workflow that enables almost everything with bpf mode. Then, rather than having a dedicated workflow for tproxy, tproxy would be just one detail of the "bpf-everything mode" workflow.

One downside to this is that any bugs specific to bpf.tproxy may be less easy to recognize when the test fails. However, I think that we should also work towards making sure that Cilium reports failures in a clear way and that the CI exposes those failure conditions clearly to developers. This path would help to mitigate such issues both in development and also potentially for users at runtime, who may not have the opportunity to just evaluate whether this particular flag on/off impacts their network.

**Comment by joestringer** [comment]
(I didn't actually review the workflow, was going for a holistic evaluation of the idea of where this testing belongs, so I re-requested that codeowner)

File: .github/workflows/conformance-bpf-tproxy.yaml
Comment (aanm) [reply .github/workflows/conformance-bpf-tproxy.yaml:1]: I have the same comments

Comment (mhofstetter) [reply .github/workflows/conformance-bpf-tproxy.yaml:1]: Thanks for the feedback @joestringer. I do understand your concerns. The main goal was to introduce some basic coverage - to prevent regression once again.

Wouldn't your proposal be covered by introducing an additional "matrix property" into the existing [Conformance E2E tests](https://github.com/cilium/cilium/blob/main/.github/workflows/conformance-e2e.yaml#L75-L113) (e.g. `tproxyMode` (`iptables` / `bpf`) - that would enable `bpf.tproxy=true/false`)? 

As far as i understand the Conformance E2E tests, there main goal is to combine workflows like the one your described (testing "everything" with feature X enabled or not) in one workflow. To avoid having to test the feature with all other possible feature combinations (IMO an overkill), we could simply introduce an additional matrix combination (`include`) where we would enable `tproxyMode=bpf`.

WDYT?

Comment (joestringer) [reply .github/workflows/conformance-bpf-tproxy.yaml:1]: That sounds like a good idea yeah. I would have a mild preference to combine the tproxy mode with one of the existing matrix combinations rather than a new matrix combination, but at the very least it will be much simpler to maintain if this change is made in the matrix to the existing workflow rather than the current proposal for adding a whole extra workflow. :+1:

**Pull request closed without merging**
